File: 1008d2/dgpt.cpp
#include <bits/stdc++.h>
using namespace std;
// Get-Content t.txt | .\dgpt.exe

void solve() {
    int t;
    cin >> t; // Number of test cases

    while (t--) {
        int n;
        cin >> n; // Number of gate pairs

        vector<pair<string, int>> leftGates(n), rightGates(n);
        for (int i = 0; i < n; ++i) {
            cin >> leftGates[i].first >> leftGates[i].second;
            cin >> rightGates[i].first >> rightGates[i].second;
        }

        // DP table: dp[l][r] stores max total after processing some gates
        map<pair<int, int>, int> dp;
        dp[{1, 1}] = 2; // Initially, 1 person in each lane

        for (int i = 0; i < n; ++i) {
            map<pair<int, int>, int> newDp;

            for (auto &[state, total] : dp) {
                int leftLane = state.first;
                int rightLane = state.second;

                // Calculate gains from current pair of gates
                int leftGain = 0, rightGain = 0;

                if (leftGates[i].first == "+") {
                    leftGain = leftGates[i].second;
                } else if (leftGates[i].first == "x") {
                    leftGain = (leftGates[i].second - 1) * leftLane;
                }

                if (rightGates[i].first == "+") {
                    rightGain = rightGates[i].second;
                } else if (rightGates[i].first == "x") {
                    rightGain = (rightGates[i].second - 1) * rightLane;
                }

                // Try all possible allocations of gains
                for (int allocateToLeft = 0; allocateToLeft <= leftGain + rightGain; ++allocateToLeft) {
                    int newLeftLane = leftLane + allocateToLeft;
                    int newRightLane = rightLane + (leftGain + rightGain - allocateToLeft);

                    newDp[{newLeftLane, newRightLane}] =
                        max(newDp[{newLeftLane, newRightLane}], total + leftGain + rightGain);
                }
            }

            dp = move(newDp); // Update DP table
        }

        // Find maximum total in DP table
        int maxTotal = 0;
        for (auto &[state, total] : dp) {
            maxTotal = max(maxTotal, total);
        }

        cout << maxTotal << endl;
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    solve();
    return 0;
}

File: 1008d2/dgpt.py
def solve():
    import sys
    input = sys.stdin.read
    data = input().splitlines()
    
    t = int(data[0])  # Number of test cases
    results = []
    idx = 1
    
    for _ in range(t):
        n = int(data[idx])  # Number of gate pairs
        idx += 1
        
        gates = []
        for _ in range(n):
            left_gate, right_gate = data[idx].split(), data[idx + 1].split()
            gates.append((left_gate, right_gate))
            idx += 2
        
        # Initial values
        left_lane = right_lane = 1
        
        for left_gate, right_gate in gates:
            new_left_lane = new_right_lane = 0
            
            # Parse left gate
            if left_gate[0] == '+':
                add_left = int(left_gate[1])
                new_left_lane += add_left
            elif left_gate[0] == 'x':
                mult_left = int(left_gate[1])
                new_left_lane += (mult_left - 1) * left_lane
            
            # Parse right gate
            if right_gate[0] == '+':
                add_right = int(right_gate[1])
                new_right_lane += add_right
            elif right_gate[0] == 'x':
                mult_right = int(right_gate[1])
                new_right_lane += (mult_right - 1) * right_lane
            
            # Allocate optimally between lanes
            total_gain = new_left_lane + new_right_lane
            left_gain = total_gain // 2
            right_gain = total_gain - left_gain
            
            left_lane += left_gain
            right_lane += right_gain
        
        results.append(left_lane + right_lane)
    
    sys.stdout.write('\n'.join(map(str, results)) + '\n')
